File: Android/controllers/download_controller.py
"""
Download Controller - Manages download operations
"""
import threading
from typing import Callable, Optional
from models.ytdlp_wrapper import YTDLPWrapper
from models.video_model import VideoInfo


class DownloadController:
    """Controller for handling download operations"""
    
    def __init__(self, download_path: str):
        """Initialize download controller"""
        self.ytdlp = YTDLPWrapper(download_path)
        self.active_downloads = {}
        self.download_history = []
    
    def download_video(
        self,
        video: VideoInfo,
        format_type: str = 'mp4',
        progress_callback: Optional[Callable[[float], None]] = None,
        complete_callback: Optional[Callable[[str], None]] = None,
        error_callback: Optional[Callable[[str], None]] = None
    ):
        """Start downloading a video in a separate thread"""
        def download_thread():
            try:
                success = self.ytdlp.download_video(
                    video.url,
                    format_type,
                    progress_callback,
                    complete_callback
                )
                
                if success:
                    self.download_history.append({
                        'video': video,
                        'format': format_type,
                        'status': 'completed'
                    })
                else:
                    if error_callback:
                        error_callback("Download failed")
            except Exception as e:
                if error_callback:
                    error_callback(str(e))
            finally:
                if video.video_id in self.active_downloads:
                    del self.active_downloads[video.video_id]
        
        thread = threading.Thread(target=download_thread, daemon=True)
        self.active_downloads[video.video_id] = thread
        thread.start()
    
    def download_playlist(
        self,
        playlist: VideoInfo,
        format_type: str = 'mp4',
        progress_callback: Optional[Callable[[float], None]] = None,
        complete_callback: Optional[Callable[[str], None]] = None,
        error_callback: Optional[Callable[[str], None]] = None
    ):
        """Start downloading a playlist in a separate thread"""
        def download_thread():
            try:
                success = self.ytdlp.download_playlist(
                    playlist.url,
                    format_type,
                    progress_callback,
                    complete_callback
                )
                
                if success:
                    self.download_history.append({
                        'video': playlist,
                        'format': format_type,
                        'status': 'completed'
                    })
                else:
                    if error_callback:
                        error_callback("Playlist download failed")
            except Exception as e:
                if error_callback:
                    error_callback(str(e))
            finally:
                if playlist.video_id in self.active_downloads:
                    del self.active_downloads[playlist.video_id]
        
        thread = threading.Thread(target=download_thread, daemon=True)
        self.active_downloads[playlist.video_id] = thread
        thread.start()
    
    def is_downloading(self, video_id: str) -> bool:
        """Check if a video is currently being downloaded"""
        return video_id in self.active_downloads

File: Android/React-Native/src/controllers/downloadController.ts
import { VideoInfo } from "../models/videoModel";
import { YTDLPWrapper } from "../services/ytdlpWrapper";

export type DownloadTask = {
  downloadId: string;
  video: VideoInfo;
  formatType: "mp4" | "mp3";
  status: "queued" | "downloading" | "paused" | "completed" | "canceled" | "error";
  progress: number;
  error?: string;
  pauseFlag?: boolean;
  cancelFlag?: boolean;
};

const createId = (): string => Math.random().toString(36).slice(2, 10);

export class DownloadController {
  private ytdlp: YTDLPWrapper;
  public allDownloads = new Map<string, DownloadTask>();
  public downloadHistory: DownloadTask[] = [];

  constructor(downloadPath: string, cachePath: string, downloadPathIsSaf?: boolean) {
    // downloadPathIsSaf indicates that downloadPath is a SAF directory URI
    this.ytdlp = new YTDLPWrapper(cachePath, downloadPath, downloadPathIsSaf);
  }

  async downloadVideo(
    video: VideoInfo,
    formatType: "mp4" | "mp3" = "mp4",
    progressCallback?: (progress: number) => void,
    completeCallback?: (filename: string) => void,
    errorCallback?: (error: string) => void
  ): Promise<string> {
    const downloadId = createId();
    const task: DownloadTask = {
      downloadId,
      video,
      formatType,
      status: "queued",
      progress: 0,
      pauseFlag: false,
      cancelFlag: false,
    };

    this.allDownloads.set(downloadId, task);

    task.status = "downloading";

    const success = await this.ytdlp.downloadVideo(
      video.url,
      formatType,
      (progress) => {
        task.progress = progress;
        progressCallback?.(progress);
      },
      (filename) => {
        if (!task.cancelFlag) {
          task.progress = 100;
          task.status = "completed";
          completeCallback?.(filename);
        }
      }
    );

    if (!success) {
      if (task.cancelFlag) {
        task.status = "canceled";
        task.error = "Canceled";
      } else {
        task.status = "error";
        task.error = "Download failed";
        errorCallback?.(task.error);
      }
    }

    this.downloadHistory.push(task);
    return downloadId;
  }

  pauseDownload(downloadId: string): boolean {
    const task = this.allDownloads.get(downloadId);
    if (!task || !["downloading", "queued"].includes(task.status)) {
      return false;
    }
    task.pauseFlag = true;
    task.status = "paused";
    return true;
  }

  resumeDownload(downloadId: string): boolean {
    const task = this.allDownloads.get(downloadId);
    if (!task || task.status !== "paused") {
      return false;
    }
    task.pauseFlag = false;
    task.status = "downloading";
    return true;
  }

  cancelDownload(downloadId: string): boolean {
    const task = this.allDownloads.get(downloadId);
    if (!task || ["completed", "canceled", "error"].includes(task.status)) {
      return false;
    }
    task.cancelFlag = true;
    task.pauseFlag = false;
    task.status = "canceled";
    task.error = "Canceled";
    return true;
  }

  getDownload(downloadId: string): DownloadTask | undefined {
    return this.allDownloads.get(downloadId);
  }

  getActiveDownloads(): DownloadTask[] {
    return Array.from(this.allDownloads.values()).filter(
      (task) => !["completed", "canceled", "error"].includes(task.status)
    );
  }

  getAllDownloads(): DownloadTask[] {
    return Array.from(this.allDownloads.values());
  }
}
